### PR TITLE
GM

### PR DESCRIPTION
**Pull request by mayuanchenma**

### CODE REVIEW

File: tools/tpm2_certify.c
Comment (idesai) [inline tools/tpm2_certify.c:258]:
```suggestion
    case 2:
```

File: tools/tpm2_certify.c
Comment (idesai) [inline tools/tpm2_certify.c:306]:
```suggestion
      { "scheme",               required_argument, NULL, 2 },
```

File: tools/tpm2_certify.c
Comment (idesai) [inline tools/tpm2_certify.c:315]:
```suggestion
    *opts = tpm2_options_new("P:p:g:o:s:c:C:f:S:", ARRAY_LEN(topts), topts,
```

File: tools/tpm2_createek.c
Comment (idesai) [inline tools/tpm2_createek.c:204]: is this specific to SM2?

File: tools/tpm2_getcommandauditdigest.c
Comment (idesai) [inline tools/tpm2_getcommandauditdigest.c:117]:
```suggestion
        { "scheme",         required_argument, NULL, 0 },
```

File: tools/tpm2_getsessionauditdigest.c
Comment (idesai) [inline tools/tpm2_getsessionauditdigest.c:98]:
```suggestion
    case 0:
```

File: tools/tpm2_quote.c
Comment (idesai) [inline tools/tpm2_quote.c:232]:
```suggestion
    case 1:
```

File: tools/tpm2_quote.c
Comment (idesai) [inline tools/tpm2_quote.c:261]:
```suggestion
        { "scheme",         required_argument, NULL, 1 },
```

File: tools/tpm2_quote.c
Comment (idesai) [inline tools/tpm2_quote.c:265]:
```suggestion
    *opts = tpm2_options_new("c:p:l:q:s:m:o:F:f:g:", ARRAY_LEN(topts), topts,
```

File: tools/tpm2_startauthsession.c
Comment (idesai) [inline tools/tpm2_startauthsession.c:87]:
```suggestion
    case 'G':
```

File: tools/tpm2_startauthsession.c
Comment (idesai) [inline tools/tpm2_startauthsession.c:142]:
```suggestion
    *opts = tpm2_options_new("g:G:S:c:", ARRAY_LEN(topts), topts, on_option,
```

File: tools/tpm2_certify.c
Comment (idesai) [inline tools/tpm2_certify.c:183]:
```suggestion
        ctx.signing_key.object.tr_handle, &ctx.halg, ctx.sig_scheme,
        &ctx.scheme);
```

**Review by idesai — CHANGES_REQUESTED** [review]

**Comment by idesai** [comment]
@mayuanchenma also can you please rebase to current master and resolve the conflicts. Thanks.

File: tools/tpm2_startauthsession.c
Comment (idesai) [inline tools/tpm2_startauthsession.c:142]:
```suggestion
    *opts = tpm2_options_new("g:G:S:c:", ARRAY_LEN(topts), topts, on_option,
```
```suggestion
    *opts = tpm2_options_new("g:G:s:S:c:", ARRAY_LEN(topts), topts, on_option,
```

Comment (idesai) [inline tools/tpm2_startauthsession.c:142]:
```suggestion
    *opts = tpm2_options_new("g:G:S:c:", ARRAY_LEN(topts), topts, on_option,
```
```suggestion
    *opts = tpm2_options_new("g:G:s:S:c:", ARRAY_LEN(topts), topts, on_option,
```

**Comment by idesai** [comment]
@mayuanchenma, this looks good. Can you please test the SM algorithms with this branch with openssl-3?

**Comment by mayuanchenma** [comment]
tpm2_createek -c /tmp/ek_ctx -G ecc_sm2 -t -u /tmp/pubkey -f PEM
test failed in openssl3
@idesai

**Comment by idesai** [comment]
> tpm2_createek -c /tmp/ek_ctx -G ecc_sm2 -t -u /tmp/pubkey -f PEM
> ERROR: Failed to create a EC public key: error:00000000:lib(0)::reason(0)
> test failed in openssl3
> @idesai

Any other debug information?

File: tools/tpm2_createek.c
Comment (williamcroberts) [reply tools/tpm2_createek.c:205]: I belioeve its all high range NV indices

**Comment by mayuanchenma** [comment]
please merge the 2 commits. 
createek failed in openssl3 will be fixed afterward. first spent time studying openssl3.
@idesai

**Comment by williamcroberts** [comment]
> please merge the 2 commits. createek failed in openssl3 will be fixed afterward. first spent time studying openssl3. @idesai

We don't merge things that cause breaks to the CI, please upload the OpenSSL3 fix and we will merge assuming code review comments are corrected.

**Comment by mayuanchenma** [comment]
the  2 commits in openssl3 are ok.
please merge the 2 commits.
@idesai

**Comment by idesai** [comment]
> the 2 commits in openssl3 are ok.
> please merge the 2 commits.
> @idesai

@mayuanchenma thank you for submitting the pull request. Specifically the changes introduced in db06dd4d3 have to be addressed. Currently, we are able to compile the code with openssl3 but we cannot actually perform SM operations with it as the code is ifdef-compiled-out for openssl3.

**Comment by mayuanchenma** [comment]
> > the 2 commits in openssl3 are ok.
> > please merge the 2 commits.
> > @idesai
> 
> @mayuanchenma thank you for submitting the pull request. Specifically the changes introduced in [db06dd4](https://github.com/tpm2-software/tpm2-tools/commit/db06dd4d3c106abfb794f4f3738ee717f76ae52e) have to be addressed. Currently, we are able to compile the code with openssl3 but we cannot actually perform SM operations with it as the code is ifdef-compiled-out for openssl3.

SM operations test in openssl3 is ok. so ifdef-compiled-out for openssl3 is right.
@idesai

**Comment by idesai** [comment]
I was able to run this on a TPM with SM algs. So this works with a caveat that the checkquote needs to be updated too.

**Pull request merged**
